ocaml/xapi/network_event_loop.ml

```diff
@@ -0,0 +1,102 @@
+(*
+ * Copyright (C) 2017 Citrix Systems Inc.
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published
+ * by the Free Software Foundation; version 2.1 only. with the special
+ * exception on linking described in file LICENSE.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *)
+
+module D=Debug.Make(struct let name="network_event_loop" end)
+open D
+
+let _watch_networks_for_nbd_changes __context ~update_firewall ~wait_after_event_seconds ~wait_after_failure_seconds =
+
+  (* We keep track of the network objects in the database using this event loop. *)
+  let classes = ["network"] in
+  (* We keep track of the interfaces that we last passed to the firewall script
+     to allow NBD traffic on them. At startup, we don't know on which
+     interfaces NBD is allowed, and we always update the firewall. *)
+  let allowed_interfaces = None in
+
+  let api_timeout = 60. in
+  let timeout = 30. +. api_timeout +. !Db_globs.master_connection_reset_timeout in
+
+  let wait_for_network_change ~token =
+    let from =
+      Helpers.call_api_functions ~__context
+        (fun rpc session_id ->
+           Client.Client.Event.from ~rpc ~session_id ~classes ~token ~timeout |> Event_types.event_from_of_rpc)
+        ~test_fn:(fun () -> Xapi_event.from ~__context ~classes ~token ~timeout |> Event_types.parse_event_from)
+    in
+    from.Event_types.token
+  in
+
+  let localhost = Helpers.get_localhost ~__context in
+
+  let rec loop ~token ~allowed_interfaces =
+    let (token, allowed_interfaces) =
+      try
+        let token = wait_for_network_change ~token in
+        let pifs = Db.Host.get_PIFs ~__context ~self:localhost in
+        let allowed_connected_networks =
+          (* We use Valid_ref_list to continue processing the list in case some network refs are null or invalid *)
+          Valid_ref_list.filter_map
+            (fun pif ->
+               let network = Db.PIF.get_network ~__context ~self:pif in
+               let purpose = Db.Network.get_purpose ~__context ~self:network in
+               if List.mem `nbd purpose || List.mem `insecure_nbd purpose then
+                 Some network
+               else
+                 None
+            )
+            pifs
+        in
+        let interfaces = List.map (fun network -> Db.Network.get_bridge ~__context ~self:network) allowed_connected_networks in
+        let needs_firewall_update = match allowed_interfaces with
+          | Some allowed_interfaces ->
+            not (Xapi_stdext_std.Listext.List.set_equiv interfaces allowed_interfaces)
```

Comment:
The problem with idempotency is that if people mess with the firewall config themselves, e.g. using the NBD firewall config script, than things will stop working. But I guess we need this now, because it's best not to update the firewall after every irrelevant network object change.

Reply:
But actually users should not change dom0's firewall. 
If xapi getting out of sync turns out to be an issue in the future, we can use the older, caching code, and we can compare the new and old snapshots to see if any relevant fields change. Then we could call the firewall unconditionally, because we'd know when we have to update it. (the code for the old, caching version of this event loop is here: https://github.com/gaborigloi/xen-api/blob/feature/CBT_network_event_caching/ocaml/xapi/network_event_loop.ml )

Reply:
There is an older version of this code, which does caching, and we can compare the new and old snapshots to see if any relevant fields change. Then we'd know when we have to update update the firewall, and would not have to make DB queries. We can revert to this version if the current one turns out to be slow due to the DB queries. (the code is here: https://github.com/gaborigloi/xen-api/blob/feature/CBT_network_event_caching/ocaml/xapi/network_event_loop.ml )